BLE_Advertising/source/main.cpp

```diff
@@ -94,60 +98,74 @@ class BatteryDemo : ble::Gap::EventHandler {
+        _adv_data_builder.clear();
+        const uint8_t _vendor_specific_data[4] = { 0xAD, 0xDE, 0xBE, 0xEF };
+        _adv_data_builder.setManufacturerSpecificData(_vendor_specific_data);
+
+        _ble.gap().setAdvertisingScanResponse(
+            ble::LEGACY_ADVERTISING_HANDLE,
+            _adv_data_builder.getAdvertisingData()
+        );
+
+        /* start advertising */
 
         error = _ble.gap().startAdvertising(ble::LEGACY_ADVERTISING_HANDLE);
 
         if (error) {
             print_error(error, "_ble.gap().startAdvertising() failed");
             return;
         }
+
+        /* we simulate battery discharging by updating it every second */
+        _event_queue.call_every(
+            1000ms,
+            [this]() {
+                update_battery_level();
+            }
+        );
     }
 
-    void update_sensor_value() {
-        _battery_level++;
-        if (_battery_level > 100) {
-            _battery_level = 20;
+    void update_battery_level()
+    {
+        if (_battery_level-- == 10) {
+            _battery_level = 100;
         }
 
-        _battery_service.updateBatteryLevel(_battery_level);
-    }
-
-    void blink(void) {
-        led1 = !led1;
-    }
+        /* update the payload with the new value */
+        ble_error_t error = _adv_data_builder.setServiceData(GattService::UUID_BATTERY_SERVICE, make_Span(&_battery_level, 1));
 
-private:
-    /* Event handler */
-
-    void onDisconnectionComplete(const ble::DisconnectionCompleteEvent&) {
-        _ble.gap().startAdvertising(ble::LEGACY_ADVERTISING_HANDLE);
-        event_queue.cancel(_update_sensor_event);
-    }
+        if (error) {
+            print_error(error, "_adv_data_builder.setServiceData() failed");
+            return;
+        }
 
-    void onConnectionComplete(const ble::ConnectionCompleteEvent&) {
-        _update_sensor_event = _event_queue.call_every(
-            1000, this, &BatteryDemo::update_sensor_value
+        /* set the new payload, we don't need to stop advertising */
+        error = _ble.gap().setAdvertisingPayload(
+            ble::LEGACY_ADVERTISING_HANDLE,
+            _adv_data_builder.getAdvertisingData()
         );
-    }
 
+        if (error) {
+            print_error(error, "_ble.gap().setAdvertisingPayload() failed");
+            return;
+        }
+    }
 
 private:
     BLE &_ble;
     events::EventQueue &_event_queue;
 
-    UUID _battery_uuid;
-
     uint8_t _battery_level;
-    BatteryService _battery_service;
-    int _update_sensor_event;
 
     uint8_t _adv_buffer[ble::LEGACY_ADVERTISING_MAX_SIZE];
     ble::AdvertisingDataBuilder _adv_data_builder;
 };
 
-/** Schedule processing of events from the BLE middleware in the event queue. */
-void schedule_ble_events(BLE::OnEventsToProcessCallbackContext *context) {
+/* Schedule processing of events from the BLE middleware in the event queue. */
+void schedule_ble_events(BLE::OnEventsToProcessCallbackContext *context)
```

Comment:
That's something that would be worth to add in the support library (not necessary for this PR).

Reply:
agree